pkg/kube/client.go

```diff
@@ -369,8 +370,14 @@ func batchPerform(infos ResourceList, fn func(*resource.Info) error, errs chan<-
 			kind = currentKind
 		}
 		wg.Add(1)
-		go func(i *resource.Info) {
-			errs <- fn(i)
+
+		go func(info *resource.Info) {
```

Comment:
I re-read the Issue, I don't think this will solve the problem. Because it seems to be used only when it is created.

Reply:
batchPerform is used by perform and that's used by Create, Delete, WatchUntilReady

It might not impact the Update action, but my issue is not with an update to the resource. My problem is that on create, kubernetes modifies the resourcequota and throws a 409 conflict. That conflict can be worked around by simply retrying the same create again.

Reply:
> batchPerform is used by perform and that's used by Create, Delete, WatchUntilReady
> 
> Update uses `Create` when a resource does not yet exist.
> 
> It might not impact an actual update action, but my issue is not with an update to the resource (at least not in my test case ...). My problem is that inside a namespace with a resourcequota, on create of a resource that's covered by the resourcequota, kubernetes modifies the resourcequota and throws a 409 conflict for the resourcequota. That conflict can be worked around by simply retrying the same create again.

I probably understand, can you add some unit tests to prove its effectiveness? please

Reply:
Yeah, that's up next but that's where I'm stuck (due to my unfamiliarity with the project). Where would be the best place to simulate a 409 Conflict message after any resource create?

From my lab tests, I add the following debug output:
~~~
// Create creates Kubernetes resources specified in the resource list.
func (c *Client) Create(resources ResourceList) (*Result, error) {
        c.Log("creating %d resource(s)", len(resources))
        createResourceLogger := func(info *resource.Info) error {
                returnErr := createResource(info)
                if returnErr != nil {
                        c.Log(
                                "FAILED Time UnixNano: " + fmt.Sprintf("%v", time.Now().UnixNano()) +
                                        " Name: " + info.Name +
                                        ", Namespace: " + info.Namespace +
                                        ", ERROR: " + returnErr.Error(),
                        )
                        errJson, errErr := json.Marshal(returnErr)
                        if errErr == nil {
                                c.Log("==> ERROR JSON: " + string(errJson))
                        } else {
                                c.Log("Could not marshal ERROR")
                        }
                }
                return returnErr
        }

        if err := perform(resources, createResourceLogger); err != nil {
                return nil, err
        }
        return &Result{Created: resources}, nil
}

/*
// Create creates Kubernetes resources specified in the resource list.
func (c *Client) Create(resources ResourceList) (*Result, error) {
        c.Log("creating %d resource(s)", len(resources))
        if err := perform(resources, createResource); err != nil {
                return nil, err
        }
        return &Result{Created: resources}, nil
}
*/
~~~

And then when I hit the issue (it is super intermittend) I get:
~~~
$ tail -f test_gitlab.log | grep 'ERROR' -B2
(...)
client.go:315: [debug] Starting delete for "gitlab-shared-secrets-1-dzw" Job
client.go:123: [debug] creating 115 resource(s)
client.go:127: [debug] FAILED Time UnixNano: 1621355710894172045 Name: gitlab-prometheus-server, Namespace: gitlab, ERROR: Operation cannot be fulfilled on resourcequotas "quota": the object has been modified; please apply your changes to the latest version and try again
client.go:135: [debug] ==> ERROR JSON: {"ErrStatus":{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Operation cannot be fulfilled on resourcequotas \"quota\": the object has been modified; please apply your changes to the latest version and try again","reason":"Conflict","details":{"name":"quota","kind":"resourcequotas"},"code":409}}
(...)
~~~

Normally, with the above, it would have failed, but with the retry code, helm recovers.